File: symfony/src/Service/Cards/CardService.php
<?php

namespace App\Service\Cards;

use App\Entity\Cards\Card;
use App\Repository\Cards\CardRepository;

class CardService
{
    public function __construct(
        private CardRepository $repository,
    ) {}

    public function create(array $data): Card
    {
        throw new \LogicException('Not implemented');
    }

    public function update(Card $entity, array $data): Card
    {
        throw new \LogicException('Not implemented');
    }

    public function ban(int $id): void
    {
        $entity = $this->repository->find($id);
        if (!$entity) throw new \RuntimeException('Card not found: ' . $id);
        $entity->ban();
        $this->repository->save($entity, flush: true);
    }

    public function unban(int $id): void
    {
        $entity = $this->repository->find($id);
        if (!$entity) throw new \RuntimeException('Card not found: ' . $id);
        $entity->unban();
        $this->repository->save($entity, flush: true);
    }

    public function restrict(int $id): void
    {
        $entity = $this->repository->find($id);
        if (!$entity) throw new \RuntimeException('Card not found: ' . $id);
        $entity->restrict();
        $this->repository->save($entity, flush: true);
    }

    public function unrestrict(int $id): void
    {
        $entity = $this->repository->find($id);
        if (!$entity) throw new \RuntimeException('Card not found: ' . $id);
        $entity->unrestrict();
        $this->repository->save($entity, flush: true);
    }

    public function calculateValue(int $id): mixed
    {
        $entity = $this->repository->find($id);
        if (!$entity) throw new \RuntimeException('Card not found: ' . $id);
        $result = $entity->calculateValue();
        $this->repository->save($entity, flush: true);
        return $result;
    }

    public function applyRarityBonus(int $id, $multiplier): mixed
    {
        $entity = $this->repository->find($id);
        if (!$entity) throw new \RuntimeException('Card not found: ' . $id);
        $result = $entity->applyRarityBonus($multiplier);
        $this->repository->save($entity, flush: true);
        return $result;
    }

    public function isLegalInFormat(int $id, $format): mixed
    {
        $entity = $this->repository->find($id);
        if (!$entity) throw new \RuntimeException('Card not found: ' . $id);
        $result = $entity->isLegalInFormat($format);
        $this->repository->save($entity, flush: true);
        return $result;
    }
}

File: rails/app/services/cards/card_service.rb
module Cards
  class CardService
    def initialize(repository = Card)
      @repository = repository
    end

    def create(attributes)
      raise NotImplementedError, "#{self.class}#create not implemented"
    end

    def update(entity, attributes)
      raise NotImplementedError, "#{self.class}#update not implemented"
    end

    def ban(id)
      instance = Card.find(id)
      instance.ban()
      instance.save!
    end

    def unban(id)
      instance = Card.find(id)
      instance.unban()
      instance.save!
    end

    def restrict(id)
      instance = Card.find(id)
      instance.restrict()
      instance.save!
    end

    def unrestrict(id)
      instance = Card.find(id)
      instance.unrestrict()
      instance.save!
    end

    def calculate_value(id)
      instance = Card.find(id)
      result = instance.calculate_value()
      instance.save!
      result
    end

    def apply_rarity_bonus(id, multiplier)
      instance = Card.find(id)
      result = instance.apply_rarity_bonus(multiplier)
      instance.save!
      result
    end

    def is_legal_in_format(id)
      instance = Card.find(id)
      result = instance.is_legal_in_format(format)
      instance.save!
      result
    end
  end
end
